source/common/upstream/cluster_manager_impl.cc

```diff
@@ -470,38 +490,41 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
     //
     // See https://github.com/envoyproxy/envoy/pull/3941 for more context.
     bool scheduled = false;
-    const auto merge_timeout =
-        PROTOBUF_GET_MS_OR_DEFAULT(cluster.info()->lbConfig(), update_merge_window, 1000);
+    const auto merge_timeout = PROTOBUF_GET_MS_OR_DEFAULT(cm_cluster.cluster().info()->lbConfig(),
+                                                          update_merge_window, 1000);
     // Remember: we only merge updates with no adds/removes — just hc/weight/metadata changes.
     const bool is_mergeable = hosts_added.empty() && hosts_removed.empty();
 
     if (merge_timeout > 0) {
       // If this is not mergeable, we should cancel any scheduled updates since
       // we'll deliver it immediately.
-      scheduled = scheduleUpdate(cluster, priority, is_mergeable, merge_timeout);
+      scheduled = scheduleUpdate(cm_cluster, priority, is_mergeable, merge_timeout);
     }
 
     // If an update was not scheduled for later, deliver it immediately.
     if (!scheduled) {
       cm_stats_.cluster_updated_.inc();
-      postThreadLocalClusterUpdate(cluster, priority, hosts_added, hosts_removed);
+      postThreadLocalClusterUpdate(
+          cm_cluster, ThreadLocalClusterUpdateParams(priority, hosts_added, hosts_removed));
     }
   });
 
-  // Finally, if the cluster has any hosts, post updates cross-thread so the per-thread load
-  // balancers are ready.
+  // Finally, post updates cross-thread so the per-thread load balancers are ready.
+  ThreadLocalClusterUpdateParams params;
   for (auto& host_set : cluster.prioritySet().hostSetsPerPriority()) {
     if (host_set->hosts().empty()) {
       continue;
     }
-    postThreadLocalClusterUpdate(cluster, host_set->priority(), host_set->hosts(), HostVector{});
+    params.per_priority_update_params_.emplace_back(host_set->priority(), host_set->hosts(),
+                                                    HostVector{});
   }
+  postThreadLocalClusterUpdate(cm_cluster, std::move(params));
```

Comment:
Do we want this call if `params` was never populated? Before, `postThreadLocalClusterUpdate()` sould only be called if we got to that point in the above loop... (not sure what the implications of posting empty params is... an unneeded wakeup everywhere?).

Reply:
Yes this is required, it's actually the crux of the fix. We combine creation with update, regardless of whether there are hosts or not. I will add more comments.

Reply:
Can you test the cm_cluster.added_or_updated_ but not flipping it?
If the cluster is added, there is no need to post the task to worker

Reply:
We could, though I'm not sure the extra complexity is worth it. I will take another look.